lib/ide/src/controller.rs

```diff
@@ -0,0 +1,426 @@
+        }
+    }
+}
+
+
+
+// =======================
+// === Text controller ===
+// =======================
+
+/// Module controller.
+pub mod module {
+    use super::*;
+
+    /// Structure uniquely identifying module location in the project.
+    /// Mappable to filesystem path.
+    #[derive(Clone,Debug,Eq,Hash,PartialEq)]
+    pub struct Location(pub String);
+    impl Location {
+        /// Obtains path (within a project context) to the file with this module.
+        pub fn to_path(&self) -> file_manager_client::Path {
+            // TODO [mwu] Extremely provisional. When multiple files support is
+            //            added, needs to be fixed, if not earlier.
+            let result = format!("./{}.luna", self.0);
+            file_manager_client::Path::new(result)
+        }
+    }
+
+    /// State data of the module controller.
+    #[derive(Clone,Debug)]
+    pub struct Data {
+        /// This module's location.
+        pub loc      : Location,
+        /// Contents of the module file.
+        pub contents : String,
+        /// Handle to the project.
+        pub parent   : project::StrongHandle,
+    }
+
+    impl Data {
+        /// Fetches the Luna code for this module using remote File Manager.
+        pub fn fetch_text(&self) -> impl Future<Output = FallibleResult<String>> {
+            let loc    = self.loc.clone();
+            let parent = self.parent.clone();
+            // TODO [mwu] When metadata support is added, they will need to be
+            //            stripped together with idmap from the source code.
+            async move {
+                parent.read_module(loc).await
+            }
+        }
+    }
+
+    make_handles!(Data);
+
+    impl StrongHandle {
+        /// Fetches the Luna code for this module using remote File Manager.
+        pub fn fetch_text(&self) -> impl Future<Output = FallibleResult<String>> {
+            self.with(|data| data.fetch_text()).flatten()
+        }
+
+        /// Receives a notification call when file with this module has been
+        /// modified by a third-party tool (like non-IDE text editor).
+        pub async fn file_externally_modified(&self) {
+            // TODO: notify underlying text/graph controllers about the changes
+            todo!()
+        }
+    }
+}
+
+
+
+// =======================
+// === Text controller ===
+// =======================
+
+/// Text controller.
+pub mod text {
+    use super::*;
+
+    /// A single set of edits. All edits use indices relative to the document's
+    /// state from before any edits being applied.
+    pub type Edits = Vec<Edit>;
+
+    /// External context for this controller (underlying controller).
+    #[derive(Clone,Debug)]
+    pub enum Context {
+        /// Controller for the Luna module that we are displaying.
+        TextFromModule(module::StrongHandle),
+        /// Controller for the project with the non-Luna file we are displaying.
```

Comment:
Please do not use product names in source code. It makes it hard to rename in the future.

Reply:
But... How are we supposed to call it here? Our Great Language source file?